File: square/src/main/java/com/raxdenstudios/square/interceptor/BaseInterceptor.java
package com.raxdenstudios.square.interceptor;

import android.app.Activity;
import android.app.Application;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.Context;
import android.util.Log;

import com.raxdenstudios.mvp.presenter.Presenter;

/**
 * Created by Ángel Gómez
 * <p>
 * This abstract class defines the basis of an interceptor.
 */
public abstract class BaseInterceptor<TCallback extends InterceptorCallback> implements Interceptor {

    private static final String TAG = BaseInterceptor.class.getSimpleName();

    protected Context mContext;
    protected TCallback mCallback;

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Application application) {
        mContext = application.getApplicationContext();
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Application application, TCallback callback) {
        mContext = application.getApplicationContext();
        setOnInterceptorCallback(callback);
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Activity activity) {
        mContext = activity;
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Activity activity, TCallback callback) {
        mContext = activity;
        setOnInterceptorCallback(callback);
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Fragment fragment) {
        mContext = fragment.getActivity();
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Fragment fragment, TCallback callback) {
        mContext = fragment.getActivity();
        setOnInterceptorCallback(callback);
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(DialogFragment dialogFragment) {
        mContext = dialogFragment.getActivity();
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(DialogFragment dialogFragment, TCallback callback) {
        mContext = dialogFragment.getActivity();
        setOnInterceptorCallback(callback);
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Presenter presenter) {
        mContext = presenter.mContext;
    }

    @SuppressWarnings("unchecked")
    public BaseInterceptor(Presenter presenter, TCallback callback) {
        mContext = presenter.mContext;
        setOnInterceptorCallback(callback);
    }

    public void setOnInterceptorCallback(TCallback callback) {
        mCallback = callback;
    }

    @Override
    public void onCreate() {
        Log.d(TAG, this.getClass().getSimpleName() + " created!");
    }

    @Override
    public void onDestroy() {
        Log.d(TAG, this.getClass().getSimpleName() + " destroyed!");
        mCallback = null;
    }

}

File: square/src/main/java/com/raxdenstudios/square/interceptor/BaseInterceptor.kt
package com.raxdenstudios.square.interceptor

import android.app.Activity
import android.app.Application
import android.content.Context
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.util.Log

/**
 * Created by Ángel Gómez
 *
 *
 * This abstract class defines the basis of an interceptor.
 */
abstract class BaseInterceptor<TCallback : InterceptorCallback>(
        protected var context: Context,
        protected var callback: TCallback? = null)
    : Interceptor {

    constructor(application: Application, callback: TCallback? = null) : this(application.applicationContext, callback)

    constructor(activity: Activity, callback: TCallback? = null) : this(activity as Context, callback)

    constructor(fragment: Fragment, callback: TCallback? = null) : this(fragment.activity as Context, callback)

    constructor(dialogFragment: DialogFragment, callback: TCallback? = null) : this(dialogFragment.activity as Context, callback)

    // =============================================================================================

    override fun onCreate() {
        Log.d(BaseInterceptor::class.java.simpleName, this.javaClass.simpleName + " created!")
    }

    override fun onDestroy() {
        Log.d(BaseInterceptor::class.java.simpleName, this.javaClass.simpleName + " destroyed!")
        callback = null
    }

}
